File: coppafisher/find_spots/detect.jl
module Detect

using Base
using NearestNeighbors
using NPZ


function detect_spots_io(args::Vector{String})::Nothing
    if length(args) != 4
        error("detect_spots_io requires four arguments")
    end

    image_filepath = args[1]
    intensity_thresh = parse(Float32, args[2])
    radius_xy = parse(Float32, args[3])
    radius_z = parse(Float32, args[4])

    if !isfile(image_filepath)
        error("No image file found at " * image_filepath)
    end

    image = npzread(image_filepath)

    result = Detect.detect_spots(image, intensity_thresh, radius_xy, radius_z)
    maxima_yxz = result[1]
    maxima_intensity = result[2]

    maxima_yxz_filepath = joinpath(dirname(image_filepath), "maxima_yxz.npy")
    maxima_intensity_filepath = joinpath(dirname(image_filepath), "maxima_intensity.npy")

    # Overwrite any pre-existing results.
    rm(maxima_yxz_filepath, force=true)
    rm(maxima_intensity_filepath, force=true)
    npzwrite(maxima_yxz_filepath, maxima_yxz)
    npzwrite(maxima_intensity_filepath, maxima_intensity)
end


function detect_spots(
    image::Array{Float32, 3},
    intensity_thresh::Number,
    radius_xy::Number,
    radius_z::Number
)::Tuple{Matrix{Int16},Vector{Float32}}
    """
    The same function as the Python version located at `coppafisher/find_spots/detect.py`.

    The only difference is that removal of duplicate spots is always on.
    """
    if ndims(image) != 3
        error("Image must be three dimensional")
    end
    if typeof(image) != Array{Float32, 3}
        error("Image must be Array{Float32, 3}")
    end
    if intensity_thresh < 0
        error("intensity_thresh must be >= 0")
    end
    if radius_xy <= 0
        error("radius_xy must be > 0")
    end
    if radius_z <= 0
        error("radius_z must be > 0")
    end

    # Gather all image pixels above intensity_thresh.
    indices = findall(x -> x > intensity_thresh, image)
    n_spots = length(indices)
    maxima_locations = Matrix{Int16}(undef, 3, n_spots)
    maxima_intensities = Vector{Float32}(undef, n_spots)
    for (i, index) in enumerate(indices)
        maxima_locations[:, i] = [index[1], index[2], index[3]]
        maxima_intensities[i] = image[index]
    end
    indices = nothing

    # Find the nearest neighbour(s) for every maxima.
    maxima_locations_norm = copy(maxima_locations)
    maxima_locations_norm = convert(Matrix{Float32}, maxima_locations_norm)
    maxima_locations_norm[3, :] .*= Float32(radius_xy / radius_z)
    maxima_locations_norm[3, :] .+= eps(Float32)
    tree = KDTree(maxima_locations_norm)
    # The Julia code captures inclusively from the radius compared to Python, so I subtract a small amount off.
    in_range_radius = Float32(radius_xy) - eps(Float32)
    pairs::Vector{Vector{Int64}} = inrange(tree, maxima_locations_norm, in_range_radius)
    maxima_locations_norm = nothing

    # Keep the most intense maxima if there are nearby neighbours.
    keep_maxima = Bool[length(pair) == 1 for pair in pairs]
    for (i, i_pairs) in enumerate(pairs)
        if any(keep_maxima[i_pairs])
            # A near neighbour has already been kept.
            continue
        end
        if all(maxima_intensities[i] .>= maxima_intensities[i_pairs])
            keep_maxima[i_pairs] .= false
            keep_maxima[i] = true
        end
    end
    maxima_locations = maxima_locations[:, keep_maxima]
    maxima_intensities = maxima_intensities[keep_maxima]

    # (3, n_spots) -> (n_spots, 3).
    maxima_locations = permutedims(maxima_locations)

    # Maxima locations indexing starts at 1! We convert it back to starting at 0 here.
    maxima_locations = broadcast(-, maxima_locations, 1)

    return (maxima_locations, maxima_intensities)
end

end

using .Detect

File: coppafisher/find_spots/detect.py
import math
from typing import List, Optional, Union

import numpy as np
import scipy
import torch

MAX_PARTITION: int = 2_000_000


def detect_spots(
    image: Union[np.ndarray, torch.Tensor] | List[Union[np.ndarray, torch.Tensor]],
    intensity_thresh: float,
    remove_duplicates: bool = False,
    radius_xy: Optional[int] = None,
    radius_z: Optional[int] = None,
) -> tuple[np.ndarray, np.ndarray]:
    """
    Spots are detected as local maxima on the given image above intensity_thresh.

    Args:
        image (`(im_y x im_x x im_z) ndarray[int or float] or tensor[int or float]` or list of one image): image to
            detect the local maxima.
        intensity_thresh (float): local maxima are greater than intensity_thresh.
        remove_duplicates (bool, optional): if two or more local maxima are close together, then only the greatest
            maxima value is detected. If they have identical intensities, one is chosen over the other. Default: false.
        radius_xy (int, optional): two local maxima are considered close together if their distance along x and/or y is
            less than radius_xy. Default: not given.
        radius_z (int, optional): two local maxima are considered close together if their distance along z is less than
            radius_xy. Default: not given.

    Returns:
        Tuple containing:
            - (`(n_spots x 3) ndarray[int16]`): maxima_yxz. The y, x, and z coordinate positions of local maxima.
            - (`(n_spots) ndarray[image.dtype]`): maxima_intensity. maxima_intensity[i] is the image intensity at maxima_yxz[i].

    Notes:
        - If the image is given as a mutable list, then the image will be popped and removed from the list once
            finished.
    """
    if type(image) is list:
        image = image.pop()
    assert type(image) is np.ndarray or type(image) is torch.Tensor
    assert type(intensity_thresh) is float
    assert type(remove_duplicates) is bool
    assert image.ndim == 3
    if remove_duplicates:
        assert radius_xy > 0
        assert radius_z > 0

    def find_locations_to_keep(locations: np.ndarray[np.int16], intensity: np.ndarray[np.float32]) -> np.ndarray[bool]:
        locations_norm = locations.copy().astype(np.float32)
        locations_norm[:, 2] *= radius_xy / radius_z
        kdtree = scipy.spatial.KDTree(locations_norm)
        # Gives a list for each maxima that contains a list of indices that are nearby neighbours, including itself.
        pairs = kdtree.query_ball_tree(kdtree, r=radius_xy)
        keep = np.array([len(pair) == 1 for pair in pairs], bool)
        for i, i_pairs in enumerate(pairs):
            if keep[i_pairs].any():
                # A near neighbour has already been kept.
                continue
            if (intensity[i] >= intensity[i_pairs]).all():
                keep[i_pairs] = False
                keep[i] = True
        return keep

    # (n_spots x 3) coordinate positions of the image local maxima.
    maxima_locations = np.array(np.array(image > intensity_thresh).nonzero()).T.astype(np.int16)
    maxima_intensities = np.array(image[tuple(maxima_locations.T)])
    del image

    if not remove_duplicates:
        return maxima_locations, maxima_intensities

    # Sometimes the KDTree query gets too large and causes a memory crash.
    # Therefore, n_spots is partitioned.
    n_partitions = math.ceil(maxima_intensities.size / MAX_PARTITION)
    maxima_location_partitions = np.array_split(maxima_locations, n_partitions)
    maxima_intensity_partitions = np.array_split(maxima_intensities, n_partitions)

    keep_maxima = np.zeros_like(maxima_intensities, bool)
    index_start = 0
    for locations_partition, maxima_intensity_partition in zip(
        maxima_location_partitions, maxima_intensity_partitions, strict=True
    ):
        keep_partition = find_locations_to_keep(locations_partition, maxima_intensity_partition)
        keep_maxima[index_start : index_start + keep_partition.size] = keep_partition
        index_start += keep_partition.size
    maxima_locations = maxima_locations[keep_maxima]
    maxima_intensities = maxima_intensities[keep_maxima]
    keep = find_locations_to_keep(maxima_locations, maxima_intensities)
    maxima_locations = maxima_locations[keep]
    maxima_intensities = maxima_intensities[keep]

    return maxima_locations, maxima_intensities
